### ISSUE TITLE
configurable_product_options_selection returns an incomplete result set

### ISSUE THREAD

**Issue by paales** [magento/magento2#38037]
### Preconditions and environment

- Magento 2.4.6

### Steps to reproduce

**1)** Get a configurable product with two super attributes and disable two simples:

1. Venia product with two super attributes: https://venia.magento.com/penelope-peasant-blouse.html
        - disable the 'Size:L, Khaki' and 'Size:L, Lilac' simple products.
2. Luma demo product with two super attributes: https://magento2-demo.magebit.com/gabrielle-micro-sleeve-top.html
        - disable the 'Size:L, Green' and 'Size:L, Red' simple products.
3. GraphCommerce demo product with two super attributes: https://graphcommerce-git-feature-gcom-1240-graphcommerce.vercel.app/nl/p/gc-puncturing-sock
        - In our own demo we've disabled 'Size:41-46, Cactus Turquoize' and 'Size41-46' Cactus Purple'

**2)** The result now should be that you aren't able to order the combination on the regular frontend.

**3)** To build the configurable product form on the product page, we query the `configurable_product_options_selection` to see which combinations are available:

(You can follow along in our demo with (expensive) workarounds applied https://graphcommerce-git-feature-gcom-1240-graphcommerce.vercel.app/nl/p/gc-puncturing-sock)

```graphql
query ConfigurableProductOptionsSelection {
  route(url: "gc-puncturing-sock") {
    ... on ConfigurableProduct {
      __typename
      sku
      ... on ConfigurableProduct {
        configurable_product_options_selection {
          options_available_for_selection {
            attribute_code
            option_value_uids
          }
        }
      }
    }
  }
}
```

We get the option_value_uids back that are available in the form. These option_value_uids determine which values should be enabled in the form and which options should be disabled:

```json
{
  "data": {
    "route": {
      "__typename": "ConfigurableProduct",
      "sku": "GC-PUNCTURING-SOCK",
      "configurable_product_options_selection": {
        "options_available_for_selection": [
          {
            "attribute_code": "size",
            "option_value_uids": [
              "Y29uZmlndXJhYmxlLzE0NC8yMTU=",
              "Y29uZmlndXJhYmxlLzE0NC8yMTY=",
              "Y29uZmlndXJhYmxlLzE0NC8yMTc=",
              "Y29uZmlndXJhYmxlLzE0NC8yMTg="
            ]
          },
          {
            "attribute_code": "print_pattern_swatch",
            "option_value_uids": [
              "Y29uZmlndXJhYmxlLzE3OC8xODAx",
              "Y29uZmlndXJhYmxlLzE3OC8xODAy",
              "Y29uZmlndXJhYmxlLzE3OC8xODA0"
            ]
          }
        ]
      }
    }
  }
}
```

In this case you can see that all options are enabled, as can be confirmed on the product page of our demo: https://graphcommerce-git-feature-gcom-1240-graphcommerce.vercel.app/nl/p/gc-puncturing-sock

**4)** Now the user selects a '4-6Y':

```graphql
query ConfigurableProductOptionsSelection {
  route(url: "gc-puncturing-sock") {
    ... on ConfigurableProduct {
      __typename
      sku
      ... on ConfigurableProduct {
       configurable_product_options_selection(
          configurableOptionValueUids: ["Y29uZmlndXJhYmxlLzE0NC8yMTU="]
        ) {
          options_available_for_selection {
            attribute_code
            option_value_uids
          }
        }
      }
    }
  }
}
```

This results in: 

```json
{
  "data": {
    "route": {
      "__typename": "ConfigurableProduct",
      "sku": "GC-PUNCTURING-SOCK",
      "configurable_product_options_selection": {
        "options_available_for_selection": [
          {
            "attribute_code": "size",
            "option_value_uids": [
              "Y29uZmlndXJhYmxlLzE0NC8yMTU="
            ]
          },
          {
            "attribute_code": "print_pattern_swatch",
            "option_value_uids": [
              "Y29uZmlndXJhYmxlLzE3OC8xODAx",
              "Y29uZmlndXJhYmxlLzE3OC8xODAy",
              "Y29uZmlndXJhYmxlLzE3OC8xODA0"
            ]
          }
        ]
      }
    }
  }
}
```

❌ This creates our first problem: We can't use the output of this query to enable/disable fields for the `size` attribute as none of the siblings are available.
➡️ We would expect the siblings to be available here, so we directly use the output of the query.

**5)** The user selects a 'Cactus Turquoise' 

```graphql
query ConfigurableProductOptionsSelection {
  route(url: "gc-puncturing-sock") {
    ... on ConfigurableProduct {
      __typename
      sku
      ... on ConfigurableProduct {
        configurable_product_options_selection(
          configurableOptionValueUids: ["Y29uZmlndXJhYmxlLzE0NC8yMTU=", "Y29uZmlndXJhYmxlLzE3OC8xODAy"]
        ) {
          options_available_for_selection {
            attribute_code
            option_value_uids
          }
          variant {
            sku
          }
        }
      }
    }
  }
}
```

Resulting in the following data:

```json
{
  "data": {
    "route": {
      "__typename": "ConfigurableProduct",
      "sku": "GC-PUNCTURING-SOCK",
      "configurable_product_options_selection": {
        "options_available_for_selection": [
          {
            "attribute_code": "size",
            "option_value_uids": [
              "Y29uZmlndXJhYmxlLzE0NC8yMTU="
            ]
          },
          {
            "attribute_code": "print_pattern_swatch",
            "option_value_uids": [
              "Y29uZmlndXJhYmxlLzE3OC8xODAy"
            ]
          }
        ],
        "variant": {
          "sku": "GC-217-SOCK-6Y"
        }
      }
    }
  }
}
```

❌ This creates our second problem: We can't use the output of this query to enable/disable fields for the `size` attribute AND we can't use the output of this query to enable/disable fields for the `print_pattern_swatch` attribute as none of the siblings are available.

However we do need this query as we need to find out if the combination has a resulting variant (indicating we're done with filling in the form)

To actually get the correct result set for `size` and `print_pattern_swatch` to know which fields need to be disabled we need to do the above query with different arguments:

```graphql
# for the variant
configurableOptionValueUids: ["Y29uZmlndXJhYmxlLzE0NC8yMTU=", "Y29uZmlndXJhYmxlLzE3OC8xODAy"]

# for the size attribute
configurableOptionValueUids: ["Y29uZmlndXJhYmxlLzE3OC8xODAy"]

# for the print_pattern_swatch
configurableOptionValueUids: ["Y29uZmlndXJhYmxlLzE0NC8yMTU="]
```

This means that we need to do two additional queries to get the correct form. This is the way the workaround is implemented and this is what you can see in the network inspector on the demo.

Even worse when there are more options like here: https://www.fietsleaseholland.nl/leasefiets/cube-touring-hybrid-one
It requires a query **for each super attribute in a configurable**.

➡️ We would expect the siblings to be available here as if that sibling didn't have a selection applied so that we can actually use the option_value_uids as, I assume, intended

### Expected result

See above

### Actual result

See above

### Additional information

I know this is a complex topic and perhaps hard to follow along and we've taken quite a while to come to this conclusion, but I can't see a way for this feature to be useful without the proposed correction.

### Triage and priority

- [ ] Severity: **S0** _- Affects critical data or functionality and leaves users without workaround._
- [X] Severity: **S1** _- Affects critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S2** _- Affects non-critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S3** _- Affects non-critical data or functionality and does not force users to employ a workaround._
- [ ] Severity: **S4** _- Affects aesthetics, professional look and feel, “quality” or “usability”._

**Comment by engcom-November** [comment]
Hello @paales,

Thank you for the report and collaboration!

As mentioned in the discription it seems you need the sibling attributes as well to enable/disable fields for that particular attribute, this is not a bug but can be considered as feature request.

Hence marking this as feature request.

Thanks.
